designs/mixins.md

```diff
@@ -0,0 +1,1075 @@
+```
+
+
+### Mixins in the JSON AST
+
+Mixins are defined in the JSON AST using the `mixins` property of structure and
+union shapes. The `mixins` property is a list of objects. To match every other
+shape target used in the AST, the object supports a single member named
+`target` which defines the absolute shape ID of a structure or union marked
+with the `smithy.api#mixin` trait.
+
+```json
+{
+    "smithy": "1.1",
+    "shapes": {
+        "smithy.example#GetCityInput": {
+            "type": "structure",
+            "mixins": [
+                {
+                    "target": "smithy.example#CityResourceInput"
+                }
+            ],
+            "members": {}
+        },
+        "smithy.example#CityResourceInput": {
+            "type": "structure",
+            "traits": {
+                "smithy.api#mixin": {}
+            },
+            "members": {
+                "cityId": {
+                    "target": "smithy.api#String",
+                    "traits": {
+                        "smithy.api#httpLabel": {},
+                        "smithy.api#required": {}
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+
+### Mixins in selectors
+
+A new relationship is introduced to selectors called `mixin` that traverses from
+structure and union shapes to every mixin applied to them. The members of each
+applied mixin are connected to the structure or union through a normal `member`
+relationship.
+
+Given the following model:
+
+```
+@mixin
+structure PaginatedInput {
+    nextToken: String,
+    pageSize: Integer
+}
+
+structure ListSomethingInput with PaginatedInput {
+    nameFilter: String
+}
+```
+
+A `mixin` relationship exists from `ListSomethingInput` to `PaginatedInput`.
+
+The following two selectors:
+
+```
+structure[id|name = ListSomethingInput] -[mixin]-> *
+```
+
+```
+structure[id|name = ListSomethingInput] > structure
+```
+
+Both yield the following shapes:
+
+- `smithy.example#PaginatedInput`
+
+A `member` relationship exists from `ListSomethingInput` to the following
+shapes, in any order:
+
+- `smithy.example#ListSomethingInput$nameFilter`
+- `smithy.example#PaginatedInput$nextToken`
+- `smithy.example#PaginatedInput$pageSize`
+
+The following two selectors:
+
+```
+structure[id|name = ListSomethingInput] > member
+```
+
+```
+structure[id|name = ListSomethingInput] -[member]-> *
+```
+
+Both yield the following shapes, in any order:
+
+- `smithy.example#ListSomethingInput$nameFilter`
+- `smithy.example#PaginatedInput$nextToken`
+- `smithy.example#PaginatedInput$pageSize`
+
+
+### Mixins and member ordering
+
+The ordering of structure and union members is important for languages like C
+that require a stable ABI. Mixins provide a deterministic member ordering.
+Members are ordered by:
+
+- Members copied from mixins come before members defined directly in the shape.
+- Members are copied from mixins based on the order in which they applied to
+  a shape.
```

Comment:
Is it the order in which they were _first_, or _last_ applied to the shape?

Reply:
First. I'll clarify.

Reply:
Mulled this over with others, and I think we should start with a predefined ordering of members based on what we already had -- mixin members come before members defined locally, mixin members are ordered based on when they are first seen, and redefining members has no effect on ordering. If this proves insufficient for ordering, then we can later consider adding traits for explicit ordering.